library/src/window_impl.cxx

```diff
@@ -1,49 +1,321 @@
+  }
+
+  static void DisplayCameraInformation(vtkCamera* cam)
+  {
+    double* position = cam->GetPosition();
+    log::info("Camera position is: ", position[0], ", ", position[1], ", ", position[2], ".");
+    double* focalPoint = cam->GetFocalPoint();
+    log::info(
+      "Camera focal point is: ", focalPoint[0], ", ", focalPoint[1], ", ", focalPoint[2], ".");
+    double* viewUp = cam->GetViewUp();
+    log::info("Camera view up is: ", viewUp[0], ", ", viewUp[1], ", ", viewUp[2], ".");
+    log::info("Camera view angle is: ", cam->GetViewAngle(), ".\n");
+  }
+
+  bool UpdateCamera(const options& options)
+  {
+    // Set the initial camera once all options
+    // have been shown as they may have an effect on it
+    if (options.getAsInt("camera-index") < 0)
+    {
+      // set a default camera from bounds using VTK method
+      this->Renderer->vtkRenderer::ResetCamera();
+
+      // use options to overwrite camera parameters
+      vtkCamera* cam = this->Renderer->GetActiveCamera();
+
+      std::vector<double> cameraPosition = options.getAsDoubleVector("camera-position");
+      if (cameraPosition.size() == 3)
+      {
+        cam->SetPosition(cameraPosition.data());
+      }
+
+      std::vector<double> cameraFocalPoint = options.getAsDoubleVector("camera-focal-point");
+      if (cameraFocalPoint.size() == 3)
+      {
+        cam->SetFocalPoint(cameraFocalPoint.data());
+      }
+
+      std::vector<double> cameraViewUp = options.getAsDoubleVector("camera-view-up");
+      if (cameraViewUp.size() == 3)
+      {
+        cam->SetViewUp(cameraViewUp.data());
+      }
+
+      double cameraViewAngle = options.getAsDouble("camera-view-angle");
+      if (cameraViewAngle != 0)
+      {
+        cam->SetViewAngle(cameraViewAngle);
+      }
+
+      cam->Azimuth(options.getAsDouble("camera-azimuth-angle"));
+      cam->Elevation(options.getAsDouble("camera-elevation-angle"));
+      cam->OrthogonalizeViewUp();
+
+      if (options.getAsBool("verbose"))
+      {
+        window_impl::internals::DisplayCameraInformation(cam);
+      }
+    }
+
+    this->Renderer->InitializeCamera();
+    return true;
+  }
+
+  vtkSmartPointer<vtkRenderWindow> RenWin;
+  vtkSmartPointer<vtkF3DRenderer> Renderer;
+  WindowType Type;
+  const options& Options;
+};
+
 //----------------------------------------------------------------------------
-window_impl::window_impl(const options& options)
-  : Options(options)
+window_impl::window_impl(const options& options, WindowType type)
+  : Internals(new window_impl::internals(options))
 {
+  this->Internals->Type = type;
+  if (type == WindowType::NO_RENDER)
+  {
+    this->Internals->RenWin = vtkSmartPointer<vtkF3DNoRenderWindow>::New();
+  }
+  else if (type == WindowType::EXTERNAL)
+  {
+    vtkNew<vtkExternalOpenGLRenderWindow> renWin;
+    renWin->AutomaticWindowPositionAndResizeOff();
+    this->Internals->RenWin = renWin;
+    this->Internals->RenWin->SetMultiSamples(0); // Disable hardware antialiasing
```

Comment:
is this needed @Meakk ?

Reply:
Not for the external window, you can remove it